File: src/lib/aws/services/cloudwatch.ts
import {
  GetMetricStatisticsCommand,
  GetMetricDataCommand,
  ListMetricsCommand,
  type GetMetricStatisticsCommandInput,
  type GetMetricDataCommandInput,
  type MetricDataQuery,
  type Dimension
} from '@aws-sdk/client-cloudwatch';
import { getCloudWatchClient } from '../clients';
import { AWS_RESOURCES } from '../config';
import { subHours, subDays } from 'date-fns';

export interface MetricQuery {
  namespace: string;
  metricName: string;
  dimensions?: Dimension[];
  stat?: string;
  period?: number;
  startTime?: Date;
  endTime?: Date;
}

export interface LambdaMetrics {
  invocations: number;
  errors: number;
  duration: number;
  throttles: number;
  concurrentExecutions: number;
}

export interface APIGatewayMetrics {
  count: number;
  latency: number;
  '4xxErrors': number;
  '5xxErrors': number;
}

export interface DynamoDBMetrics {
  consumedReadCapacity: number;
  consumedWriteCapacity: number;
  throttledRequests: number;
  userErrors: number;
  systemErrors: number;
}

// Get metrics for a specific Lambda function
export const getLambdaMetrics = async (
  functionName: string,
  timeRange: '1h' | '6h' | '24h' | '7d' = '24h'
): Promise<LambdaMetrics> => {
  const client = getCloudWatchClient();

  const endTime = new Date();
  let startTime: Date;
  let period = 300; // 5 minutes default

  switch (timeRange) {
    case '1h':
      startTime = subHours(endTime, 1);
      period = 60; // 1 minute
      break;
    case '6h':
      startTime = subHours(endTime, 6);
      period = 300; // 5 minutes
      break;
    case '24h':
      startTime = subHours(endTime, 24);
      period = 900; // 15 minutes
      break;
    case '7d':
      startTime = subDays(endTime, 7);
      period = 3600; // 1 hour
      break;
  }

  const queries: MetricDataQuery[] = [
    {
      Id: 'm1',
      MetricStat: {
        Metric: {
          Namespace: 'AWS/Lambda',
          MetricName: 'Invocations',
          Dimensions: [{ Name: 'FunctionName', Value: functionName }]
        },
        Period: period,
        Stat: 'Sum'
      }
    },
    {
      Id: 'm2',
      MetricStat: {
        Metric: {
          Namespace: 'AWS/Lambda',
          MetricName: 'Errors',
          Dimensions: [{ Name: 'FunctionName', Value: functionName }]
        },
        Period: period,
        Stat: 'Sum'
      }
    },
    {
      Id: 'm3',
      MetricStat: {
        Metric: {
          Namespace: 'AWS/Lambda',
          MetricName: 'Duration',
          Dimensions: [{ Name: 'FunctionName', Value: functionName }]
        },
        Period: period,
        Stat: 'Average'
      }
    },
    {
      Id: 'm4',
      MetricStat: {
        Metric: {
          Namespace: 'AWS/Lambda',
          MetricName: 'Throttles',
          Dimensions: [{ Name: 'FunctionName', Value: functionName }]
        },
        Period: period,
        Stat: 'Sum'
      }
    },
    {
      Id: 'm5',
      MetricStat: {
        Metric: {
          Namespace: 'AWS/Lambda',
          MetricName: 'ConcurrentExecutions',
          Dimensions: [{ Name: 'FunctionName', Value: functionName }]
        },
        Period: period,
        Stat: 'Maximum'
      }
    }
  ];

  const input: GetMetricDataCommandInput = {
    MetricDataQueries: queries,
    StartTime: startTime,
    EndTime: endTime
  };

  const command = new GetMetricDataCommand(input);
  const response = await client.send(command);

  const metrics: LambdaMetrics = {
    invocations: 0,
    errors: 0,
    duration: 0,
    throttles: 0,
    concurrentExecutions: 0
  };

  if (response.MetricDataResults) {
    response.MetricDataResults.forEach(result => {
      const values = result.Values || [];
      const sum = values.reduce((acc, val) => acc + (val || 0), 0);
      const avg = values.length > 0 ? sum / values.length : 0;
      const max = values.length > 0 ? Math.max(...values) : 0;

      switch (result.Id) {
        case 'm1':
          metrics.invocations = sum;
          break;
        case 'm2':
          metrics.errors = sum;
          break;
        case 'm3':
          metrics.duration = avg;
          break;
        case 'm4':
          metrics.throttles = sum;
          break;
        case 'm5':
          metrics.concurrentExecutions = max;
          break;
      }
    });
  }

  return metrics;
};

// Get metrics for all Lambda functions
export const getAllLambdaMetrics = async (
  timeRange: '1h' | '6h' | '24h' | '7d' = '24h'
): Promise<Record<string, LambdaMetrics>> => {
  const results: Record<string, LambdaMetrics> = {};

  await Promise.all(
    AWS_RESOURCES.lambda.functions.map(async (functionName) => {
      try {
        results[functionName] = await getLambdaMetrics(functionName, timeRange);
      } catch (error) {
        console.error(`Error fetching metrics for ${functionName}:`, error);
        results[functionName] = {
          invocations: 0,
          errors: 0,
          duration: 0,
          throttles: 0,
          concurrentExecutions: 0
        };
      }
    })
  );

  return results;
};

// Get API Gateway metrics
export const getAPIGatewayMetrics = async (
  apiName: string,
  timeRange: '1h' | '6h' | '24h' | '7d' = '24h'
): Promise<APIGatewayMetrics> => {
  const client = getCloudWatchClient();

  const endTime = new Date();
  let startTime: Date;
  let period = 300;

  switch (timeRange) {
    case '1h':
      startTime = subHours(endTime, 1);
      period = 60;
      break;
    case '6h':
      startTime = subHours(endTime, 6);
      period = 300;
      break;
    case '24h':
      startTime = subHours(endTime, 24);
      period = 900;
      break;
    case '7d':
      startTime = subDays(endTime, 7);
      period = 3600;
      break;
  }

  const queries: MetricDataQuery[] = [
    {
      Id: 'm1',
      MetricStat: {
        Metric: {
          Namespace: 'AWS/ApiGateway',
          MetricName: 'Count',
          Dimensions: [{ Name: 'ApiName', Value: apiName }]
        },
        Period: period,
        Stat: 'Sum'
      }
    },
    {
      Id: 'm2',
      MetricStat: {
        Metric: {
          Namespace: 'AWS/ApiGateway',
          MetricName: 'Latency',
          Dimensions: [{ Name: 'ApiName', Value: apiName }]
        },
        Period: period,
        Stat: 'Average'
      }
    },
    {
      Id: 'm3',
      MetricStat: {
        Metric: {
          Namespace: 'AWS/ApiGateway',
          MetricName: '4XXError',
          Dimensions: [{ Name: 'ApiName', Value: apiName }]
        },
        Period: period,
        Stat: 'Sum'
      }
    },
    {
      Id: 'm4',
      MetricStat: {
        Metric: {
          Namespace: 'AWS/ApiGateway',
          MetricName: '5XXError',
          Dimensions: [{ Name: 'ApiName', Value: apiName }]
        },
        Period: period,
        Stat: 'Sum'
      }
    }
  ];

  const input: GetMetricDataCommandInput = {
    MetricDataQueries: queries,
    StartTime: startTime,
    EndTime: endTime
  };

  const command = new GetMetricDataCommand(input);
  const response = await client.send(command);

  const metrics: APIGatewayMetrics = {
    count: 0,
    latency: 0,
    '4xxErrors': 0,
    '5xxErrors': 0
  };

  if (response.MetricDataResults) {
    response.MetricDataResults.forEach(result => {
      const values = result.Values || [];
      const sum = values.reduce((acc, val) => acc + (val || 0), 0);
      const avg = values.length > 0 ? sum / values.length : 0;

      switch (result.Id) {
        case 'm1':
          metrics.count = sum;
          break;
        case 'm2':
          metrics.latency = avg;
          break;
        case 'm3':
          metrics['4xxErrors'] = sum;
          break;
        case 'm4':
          metrics['5xxErrors'] = sum;
          break;
      }
    });
  }

  return metrics;
};

// Get DynamoDB metrics for a specific table
export const getDynamoDBMetrics = async (
  tableName: string,
  timeRange: '1h' | '6h' | '24h' | '7d' = '24h'
): Promise<DynamoDBMetrics> => {
  const client = getCloudWatchClient();

  const endTime = new Date();
  let startTime: Date;
  let period = 300;

  switch (timeRange) {
    case '1h':
      startTime = subHours(endTime, 1);
      period = 60;
      break;
    case '6h':
      startTime = subHours(endTime, 6);
      period = 300;
      break;
    case '24h':
      startTime = subHours(endTime, 24);
      period = 900;
      break;
    case '7d':
      startTime = subDays(endTime, 7);
      period = 3600;
      break;
  }

  const queries: MetricDataQuery[] = [
    {
      Id: 'm1',
      MetricStat: {
        Metric: {
          Namespace: 'AWS/DynamoDB',
          MetricName: 'ConsumedReadCapacityUnits',
          Dimensions: [{ Name: 'TableName', Value: tableName }]
        },
        Period: period,
        Stat: 'Sum'
      }
    },
    {
      Id: 'm2',
      MetricStat: {
        Metric: {
          Namespace: 'AWS/DynamoDB',
          MetricName: 'ConsumedWriteCapacityUnits',
          Dimensions: [{ Name: 'TableName', Value: tableName }]
        },
        Period: period,
        Stat: 'Sum'
      }
    },
    {
      Id: 'm3',
      MetricStat: {
        Metric: {
          Namespace: 'AWS/DynamoDB',
          MetricName: 'ThrottledRequests',
          Dimensions: [{ Name: 'TableName', Value: tableName }]
        },
        Period: period,
        Stat: 'Sum'
      }
    },
    {
      Id: 'm4',
      MetricStat: {
        Metric: {
          Namespace: 'AWS/DynamoDB',
          MetricName: 'UserErrors',
          Dimensions: [{ Name: 'TableName', Value: tableName }]
        },
        Period: period,
        Stat: 'Sum'
      }
    },
    {
      Id: 'm5',
      MetricStat: {
        Metric: {
          Namespace: 'AWS/DynamoDB',
          MetricName: 'SystemErrors',
          Dimensions: [{ Name: 'TableName', Value: tableName }]
        },
        Period: period,
        Stat: 'Sum'
      }
    }
  ];

  const input: GetMetricDataCommandInput = {
    MetricDataQueries: queries,
    StartTime: startTime,
    EndTime: endTime
  };

  const command = new GetMetricDataCommand(input);
  const response = await client.send(command);

  const metrics: DynamoDBMetrics = {
    consumedReadCapacity: 0,
    consumedWriteCapacity: 0,
    throttledRequests: 0,
    userErrors: 0,
    systemErrors: 0
  };

  if (response.MetricDataResults) {
    response.MetricDataResults.forEach(result => {
      const values = result.Values || [];
      const sum = values.reduce((acc, val) => acc + (val || 0), 0);

      switch (result.Id) {
        case 'm1':
          metrics.consumedReadCapacity = sum;
          break;
        case 'm2':
          metrics.consumedWriteCapacity = sum;
          break;
        case 'm3':
          metrics.throttledRequests = sum;
          break;
        case 'm4':
          metrics.userErrors = sum;
          break;
        case 'm5':
          metrics.systemErrors = sum;
          break;
      }
    });
  }

  return metrics;
};

// Get metrics for all DynamoDB tables
export const getAllDynamoDBMetrics = async (
  timeRange: '1h' | '6h' | '24h' | '7d' = '24h'
): Promise<Record<string, DynamoDBMetrics>> => {
  const results: Record<string, DynamoDBMetrics> = {};

  await Promise.all(
    AWS_RESOURCES.dynamodb.tables.map(async (tableName) => {
      try {
        results[tableName] = await getDynamoDBMetrics(tableName, timeRange);
      } catch (error) {
        console.error(`Error fetching metrics for ${tableName}:`, error);
        results[tableName] = {
          consumedReadCapacity: 0,
          consumedWriteCapacity: 0,
          throttledRequests: 0,
          userErrors: 0,
          systemErrors: 0
        };
      }
    })
  );

  return results;
};

File: packages/backend/internal/aws/cloudwatch.go
package aws

import (
	"context"
	"fmt"
	"time"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/cloudwatch"
	"github.com/aws/aws-sdk-go-v2/service/cloudwatch/types"
)

// CloudWatchClient wraps the CloudWatch client
type CloudWatchClient struct {
	client *cloudwatch.Client
}

// NewCloudWatchClient creates a new CloudWatch client
func NewCloudWatchClient(cfg aws.Config) *CloudWatchClient {
	return &CloudWatchClient{
		client: cloudwatch.NewFromConfig(cfg),
	}
}

// LambdaMetrics represents Lambda function metrics
type LambdaMetrics struct {
	FunctionName string                 `json:"functionName"`
	Invocations  float64                `json:"invocations"`
	Errors       float64                `json:"errors"`
	Duration     float64                `json:"duration"`
	Throttles    float64                `json:"throttles"`
	ConcurrentExecutions float64        `json:"concurrentExecutions"`
	Period       string                 `json:"period"`
	Datapoints   []MetricDatapoint      `json:"datapoints"`
}

// MetricDatapoint represents a single metric data point
type MetricDatapoint struct {
	Timestamp time.Time `json:"timestamp"`
	Value     float64   `json:"value"`
	Unit      string    `json:"unit"`
}

// GetLambdaMetrics retrieves metrics for a Lambda function
func (c *CloudWatchClient) GetLambdaMetrics(ctx context.Context, functionName string, startTime, endTime time.Time) (*LambdaMetrics, error) {
	metrics := &LambdaMetrics{
		FunctionName: functionName,
		Period:       fmt.Sprintf("%s to %s", startTime.Format(time.RFC3339), endTime.Format(time.RFC3339)),
	}

	// Define metric queries
	queries := []types.MetricDataQuery{
		{
			Id: aws.String("invocations"),
			MetricStat: &types.MetricStat{
				Metric: &types.Metric{
					Namespace:  aws.String("AWS/Lambda"),
					MetricName: aws.String("Invocations"),
					Dimensions: []types.Dimension{
						{
							Name:  aws.String("FunctionName"),
							Value: aws.String(functionName),
						},
					},
				},
				Period: aws.Int32(300), // 5 minutes
				Stat:   aws.String("Sum"),
			},
			ReturnData: aws.Bool(true),
		},
		{
			Id: aws.String("errors"),
			MetricStat: &types.MetricStat{
				Metric: &types.Metric{
					Namespace:  aws.String("AWS/Lambda"),
					MetricName: aws.String("Errors"),
					Dimensions: []types.Dimension{
						{
							Name:  aws.String("FunctionName"),
							Value: aws.String(functionName),
						},
					},
				},
				Period: aws.Int32(300),
				Stat:   aws.String("Sum"),
			},
			ReturnData: aws.Bool(true),
		},
		{
			Id: aws.String("duration"),
			MetricStat: &types.MetricStat{
				Metric: &types.Metric{
					Namespace:  aws.String("AWS/Lambda"),
					MetricName: aws.String("Duration"),
					Dimensions: []types.Dimension{
						{
							Name:  aws.String("FunctionName"),
							Value: aws.String(functionName),
						},
					},
				},
				Period: aws.Int32(300),
				Stat:   aws.String("Average"),
			},
			ReturnData: aws.Bool(true),
		},
		{
			Id: aws.String("throttles"),
			MetricStat: &types.MetricStat{
				Metric: &types.Metric{
					Namespace:  aws.String("AWS/Lambda"),
					MetricName: aws.String("Throttles"),
					Dimensions: []types.Dimension{
						{
							Name:  aws.String("FunctionName"),
							Value: aws.String(functionName),
						},
					},
				},
				Period: aws.Int32(300),
				Stat:   aws.String("Sum"),
			},
			ReturnData: aws.Bool(true),
		},
		{
			Id: aws.String("concurrent"),
			MetricStat: &types.MetricStat{
				Metric: &types.Metric{
					Namespace:  aws.String("AWS/Lambda"),
					MetricName: aws.String("ConcurrentExecutions"),
					Dimensions: []types.Dimension{
						{
							Name:  aws.String("FunctionName"),
							Value: aws.String(functionName),
						},
					},
				},
				Period: aws.Int32(300),
				Stat:   aws.String("Maximum"),
			},
			ReturnData: aws.Bool(true),
		},
	}

	// Get metric data
	input := &cloudwatch.GetMetricDataInput{
		MetricDataQueries: queries,
		StartTime:        &startTime,
		EndTime:          &endTime,
	}

	result, err := c.client.GetMetricData(ctx, input)
	if err != nil {
		return nil, fmt.Errorf("failed to get metric data: %w", err)
	}

	// Process results
	for _, metricResult := range result.MetricDataResults {
		if metricResult.Id == nil || len(metricResult.Values) == 0 {
			continue
		}

		// Calculate sum of all values for aggregated metrics
		var total float64
		for _, value := range metricResult.Values {
			total += value
		}

		// For duration, we want the average across all data points
		if *metricResult.Id == "duration" && len(metricResult.Values) > 0 {
			total = total / float64(len(metricResult.Values))
		}

		switch *metricResult.Id {
		case "invocations":
			metrics.Invocations = total
		case "errors":
			metrics.Errors = total
		case "duration":
			metrics.Duration = total
		case "throttles":
			metrics.Throttles = total
		case "concurrent":
			// For concurrent executions, we want the maximum value
			maxConcurrent := float64(0)
			for _, value := range metricResult.Values {
				if value > maxConcurrent {
					maxConcurrent = value
				}
			}
			metrics.ConcurrentExecutions = maxConcurrent
		}

		// Add datapoints for time series (only for invocations to avoid duplication)
		if *metricResult.Id == "invocations" {
			for i, timestamp := range metricResult.Timestamps {
				if i < len(metricResult.Values) {
					metrics.Datapoints = append(metrics.Datapoints, MetricDatapoint{
						Timestamp: timestamp,
						Value:     metricResult.Values[i],
						Unit:      "Count",
					})
				}
			}
		}
	}

	return metrics, nil
}

// APIGatewayMetrics represents API Gateway metrics
type APIGatewayMetrics struct {
	APIName      string              `json:"apiName"`
	Count        float64             `json:"count"`
	Latency      float64             `json:"latency"`
	Error4XX     float64             `json:"error4xx"`
	Error5XX     float64             `json:"error5xx"`
	Period       string              `json:"period"`
	Datapoints   []MetricDatapoint   `json:"datapoints"`
}

// GetAPIGatewayMetrics retrieves metrics for an API Gateway
func (c *CloudWatchClient) GetAPIGatewayMetrics(ctx context.Context, apiName string, startTime, endTime time.Time) (*APIGatewayMetrics, error) {
	metrics := &APIGatewayMetrics{
		APIName: apiName,
		Period:  fmt.Sprintf("%s to %s", startTime.Format(time.RFC3339), endTime.Format(time.RFC3339)),
	}

	// Define metric queries
	queries := []types.MetricDataQuery{
		{
			Id: aws.String("count"),
			MetricStat: &types.MetricStat{
				Metric: &types.Metric{
					Namespace:  aws.String("AWS/ApiGateway"),
					MetricName: aws.String("Count"),
					Dimensions: []types.Dimension{
						{
							Name:  aws.String("ApiName"),
							Value: aws.String(apiName),
						},
					},
				},
				Period: aws.Int32(300),
				Stat:   aws.String("Sum"),
			},
			ReturnData: aws.Bool(true),
		},
		{
			Id: aws.String("latency"),
			MetricStat: &types.MetricStat{
				Metric: &types.Metric{
					Namespace:  aws.String("AWS/ApiGateway"),
					MetricName: aws.String("Latency"),
					Dimensions: []types.Dimension{
						{
							Name:  aws.String("ApiName"),
							Value: aws.String(apiName),
						},
					},
				},
				Period: aws.Int32(300),
				Stat:   aws.String("Average"),
			},
			ReturnData: aws.Bool(true),
		},
		{
			Id: aws.String("error4xx"),
			MetricStat: &types.MetricStat{
				Metric: &types.Metric{
					Namespace:  aws.String("AWS/ApiGateway"),
					MetricName: aws.String("4XXError"),
					Dimensions: []types.Dimension{
						{
							Name:  aws.String("ApiName"),
							Value: aws.String(apiName),
						},
					},
				},
				Period: aws.Int32(300),
				Stat:   aws.String("Sum"),
			},
			ReturnData: aws.Bool(true),
		},
		{
			Id: aws.String("error5xx"),
			MetricStat: &types.MetricStat{
				Metric: &types.Metric{
					Namespace:  aws.String("AWS/ApiGateway"),
					MetricName: aws.String("5XXError"),
					Dimensions: []types.Dimension{
						{
							Name:  aws.String("ApiName"),
							Value: aws.String(apiName),
						},
					},
				},
				Period: aws.Int32(300),
				Stat:   aws.String("Sum"),
			},
			ReturnData: aws.Bool(true),
		},
	}

	// Get metric data
	input := &cloudwatch.GetMetricDataInput{
		MetricDataQueries: queries,
		StartTime:        &startTime,
		EndTime:          &endTime,
	}

	result, err := c.client.GetMetricData(ctx, input)
	if err != nil {
		return nil, fmt.Errorf("failed to get API Gateway metrics: %w", err)
	}

	// Process results
	for _, metricResult := range result.MetricDataResults {
		if metricResult.Id == nil || len(metricResult.Values) == 0 {
			continue
		}

		// Calculate sum of all values for count metrics
		var total float64
		for _, value := range metricResult.Values {
			total += value
		}

		// For latency, we want the average across all data points
		if *metricResult.Id == "latency" && len(metricResult.Values) > 0 {
			total = total / float64(len(metricResult.Values))
		}

		switch *metricResult.Id {
		case "count":
			metrics.Count = total
		case "latency":
			metrics.Latency = total
		case "error4xx":
			metrics.Error4XX = total
		case "error5xx":
			metrics.Error5XX = total
		}

		// Add datapoints for time series (only for count to avoid duplication)
		if *metricResult.Id == "count" {
			for i, timestamp := range metricResult.Timestamps {
				if i < len(metricResult.Values) {
					metrics.Datapoints = append(metrics.Datapoints, MetricDatapoint{
						Timestamp: timestamp,
						Value:     metricResult.Values[i],
						Unit:      "Count",
					})
				}
			}
		}
	}

	return metrics, nil
}